src/Lib.P2P/Routing/DhtService.cs

```diff
@@ -0,0 +1,444 @@
+    ///   DHT Protocol version 1.0
+    /// </summary>
+    public class DhtService : IDhtService
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(DhtService));
+
+        /// <inheritdoc />
+        public virtual string Name { get; } = "libp2p-cs/kad";
+
+        /// <inheritdoc />
+        public SemVersion Version { get; } = new SemVersion(1);
+
+        /// <summary>
+        ///   Provides access to other peers.
+        /// </summary>
+        public ISwarmService SwarmService { get; set; }
+
+        /// <summary>
+        ///  Routing information on peers.
+        /// </summary>
+        public RoutingTable RoutingTable;
+
+        /// <summary>
+        ///   Peers that can provide some content.
+        /// </summary>
+        public ContentRouter ContentRouter;
+
+        /// <summary>
+        ///   The number of closer peers to return.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 20.
+        /// </value>
+        public int CloserPeerCount { get; set; } = 20;
+
+        /// <summary>
+        ///   Raised when the DHT is stopped.
+        /// </summary>
+        /// <seealso cref="StopAsync"/>
+        public event EventHandler Stopped;
+
+        /// <inheritdoc />
+        public override string ToString() { return $"/{Name}/{Version}"; }
+
+        /// <inheritdoc />
+        public async Task ProcessMessageAsync(PeerConnection connection,
+            Stream stream,
+            CancellationToken cancel = default)
+        {
+            while (true)
+            {
+                var request = await ProtoBufHelper.ReadMessageAsync<DhtMessage>(stream, cancel).ConfigureAwait(false);
+
+                _log.Debug($"got {request.Type} from {connection.RemotePeer}");
+                var response = new DhtMessage
+                {
+                    Type = request.Type,
+                    ClusterLevelRaw = request.ClusterLevelRaw
+                };
+                switch (request.Type) 
+                {
+                    case MessageType.Ping:
+                        response = ProcessPing(request, response);
+                        break;
+                    case MessageType.FindNode:
+                        response = ProcessFindNode(request, response);
+                        break;
+                    case MessageType.GetProviders:
+                        response = ProcessGetProviders(request, response);
+                        break;
+                    case MessageType.AddProvider:
+                        response = ProcessAddProvider(connection.RemotePeer, request, response);
+                        break;
+                    case MessageType.PutValue:
+                        break;
+                    case MessageType.GetValue:
+                        break;
+                    default:
+                        _log.Debug($"unknown {request.Type} from {connection.RemotePeer}");
+
+                        // TODO: Should we close the stream?
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)